service/s3/internal/customizations/response_error_middleware.go

```diff
@@ -0,0 +1,80 @@
+package customizations
+
+import (
+	"context"
+
+	awshttp "github.com/aws/aws-sdk-go-v2/aws/transport/http"
+	"github.com/awslabs/smithy-go/middleware"
+	smithyhttp "github.com/awslabs/smithy-go/transport/http"
+
+	s3http "github.com/aws/aws-sdk-go-v2/service/s3/transport/http"
+)
+
+// AddResponseErrorWrapper adds response error wrapper middleware
+func AddResponseErrorWrapper(stack *middleware.Stack) {
+	// add error wrapper middleware before operation deserializers so that it can wrap the error response
+	// returned by operation deserializers
+	stack.Deserialize.Insert(&errorWrapperMiddleware{}, "OperationDeserializer", middleware.Before)
+}
+
+type errorWrapperMiddleware struct {
+}
+
+// ID returns the middleware identifier
+func (m *errorWrapperMiddleware) ID() string {
+	return "AWSResponseErrorWrapperMiddleware"
+}
+
+func (m *errorWrapperMiddleware) HandleDeserialize(ctx context.Context, in middleware.DeserializeInput, next middleware.DeserializeHandler) (
+	out middleware.DeserializeOutput, metadata middleware.Metadata, err error,
+) {
+	out, metadata, err = next.HandleDeserialize(ctx, in)
+	if err == nil {
+		// Nothing to do when there is no error.
+		return out, metadata, err
+	}
+
+	resp, ok := out.RawResponse.(*smithyhttp.Response)
+	if !ok {
+		// No raw response to wrap with.
+		return out, metadata, err
+	}
+
+	// look for request id in metadata
+	reqID := awshttp.GetRequestIDMetadata(metadata)
+	if len(reqID) == 0 {
+		// check for request id in header `X-Amzn-Requestid`
+		if v := resp.Header.Get("X-Amzn-Requestid"); len(v) != 0 {
+			reqID = v
+			// set reqID on metadata for successful responses.
+			awshttp.SetRequestIDMetadata(&metadata, reqID)
+		}
+	}
```

Comment:
Same as generic response error wrapper, this shouldn't duplicate the request ID.